### PR TITLE
`azurerm_key_vault`: use Resource Graph API to search keyvault resource from keyvault name

### PR DESCRIPTION
**Pull request by wuxu92**
Currently the resourceClient.List API is low performance and with high probability of rasing a `context deadline exceeded` if the subscription has over thousands of resources. 

As there are several PRs try to migrate this logic to ARG before (like #13719 but closed at last), this PR is only use ARG as a high pri option and it will fall back to origin logic when ARG fails. And this PR is not to fix the issue of keyvault eventual consistant, but only to improve the performance of the `KeyVaultIDFromBaseUrl` function, Because this function is prone to causing the `context deadline exceed` error on the production environment of our customers.

To avoid breaking the `KeyVaultIDFromBaseUrl` signature, I placed the ResourceGraph Client under `resourcesClient.Client`. This way, we won't have to make changes to multiple files that depend on it.

This should be a long-term fix to #23404.

### CODE REVIEW

File: internal/services/keyvault/client/helpers.go [hunk 1]
```diff
@@ -111,6 +113,32 @@ func (c *Client) Exists(ctx context.Context, keyVaultId commonids.KeyVaultId) (b
...
+	req := graph.QueryRequest{
+		Query: fmt.Sprintf("where type =~ 'Microsoft.KeyVault/vaults' and name =~ '%s'", name),
+	}
+	resp, err := client.Resources(ctx, req)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.Model != nil && resp.Model.Data != nil {
+		if list, ok := resp.Model.Data.([]interface{}); ok && len(list) == 1 {
```
Comment (WodansSon) [inline internal/services/keyvault/client/helpers.go:129]: What if it returns `0` or `2`? Is returning the `no such keyvault in resource graph by query` the correct behavior here?

File: internal/services/keyvault/client/helpers.go [hunk 2]
```diff
@@ -111,6 +113,32 @@ func (c *Client) Exists(ctx context.Context, keyVaultId commonids.KeyVaultId) (b
...
+			if res, ok := list[0].(map[string]interface{}); ok {
+				if id, ok := res["id"]; ok {
+					if idStr, ok := id.(string); ok {
+						return idStr, nil
```
Comment (WodansSon) [inline internal/services/keyvault/client/helpers.go:133]:
```suggestion
						return pointer.To(idStr)
```

**Review by WodansSon — CHANGES_REQUESTED** [review]
Hi @wuxu92, This is looking really good so far, however I have left a few comments and suggestions on the code. It LGTM more or less, we will have to see if @jackofallops and @tombuildsstuff agrees. 🚀

File: internal/services/keyvault/client/helpers.go [hunk 3]
```diff
@@ -111,6 +113,32 @@ func (c *Client) Exists(ctx context.Context, keyVaultId commonids.KeyVaultId) (b
 	return true, nil
 }
 
+func (c *Client) KeyVaultIDFromBaseURLByGraph(ctx context.Context, client *graph.ResourcesClient, name string) (string, interface{}) {
```
[the remark this answers is not in the document]
Comment (wuxu92) [reply internal/services/keyvault/client/helpers.go]: I think `(string, error)` should be fine. I made a mistake here to define the second value as `interface{}`.

File: internal/services/keyvault/client/helpers.go [hunk 4]
```diff
@@ -130,6 +158,15 @@ func (c *Client) KeyVaultIDFromBaseUrl(ctx context.Context, resourcesClient *res
 		return &v.keyVaultId, nil
 	}
 
+	// try to get the keyVault id by resource graph API first to void the expensive ResourcesClient.List operation
+	// if it cannot fetch the keyvault from resource graph API, it will fall back to the resourceClient.List method
+	if idStr, err := c.KeyVaultIDFromBaseURLByGraph(ctx, resourcesClient.GraphResourceClient, *keyVaultName); err == nil && idStr != "" {
```
[the remark this answers is not in the document]
Comment (wuxu92) [reply internal/services/keyvault/client/helpers.go:163]: There shoud be `1` or `0` resource for keyvault resource of the same name.

File: internal/services/keyvault/client/helpers.go (hunk 1 above)
Comment (wuxu92) [reply internal/services/keyvault/client/helpers.go:129]: There shoud be `1` or `0` resource for keyvault resource of the same name.

**Comment by tombuildsstuff** [comment]
hey @wuxu92

Thanks for this PR - apologies for the delayed review here.

Whilst we've investigated the `ResourceGraph` API in the past we've noticed that it contains similar caching issues to the `Resources` API - and as such we believe that switching to load data from this API won't fundamentally resolve this issue.

That said, as a part of reviewing both the background for this PR, this PR and reviewing what's left to migrate off of `Azure/azure-sdk-for-go` more broadly - I've taken a closer look into the `Resources` API usage across the Provider and tried to determine if there's an alternate solution we can use here.

As a part of that I've reviewed the caching setup and realised that we can take a different approach with this one, rather than loading the Key Vaults on demand when populating the cache (one at a time) - since we're hitting the Azure API to get the information for the specified Key Vault - we might as well pull the entire list into the cache at the same time prior to filtering the cache for the Key Vault in question.

This means that we can use the [more up to date `ListBySubscriptions` API within the `Microsoft.KeyVault` Resource Provider](https://learn.microsoft.com/en-us/rest/api/keyvault/keyvault/vaults/list-by-subscription?view=rest-keyvault-keyvault-2022-07-01&tabs=HTTP) - which (whilst from memory) does have a lower rate limit, since we're now populating all of the Key Vaults within the Subscription - should be fine/resolve this issue. It's worth noting that this matches the approach used for Storage Accounts - but additional information can be found [in this issue](https://github.com/hashicorp/terraform-provider-azurerm/pull/24019).

As such whilst I'd like to thank you for this contribution I'm going to close this PR for the moment, as I believe that #24019 will solve #23404 (and the other issues) albeit in a different manner.

Thanks!

**Pull request closed without merging**